Fix creataing invoice originating for custom module object

indelog opened the pull request:
This fix the invoice create form when we provide `origin` element in
url. The actual create form overide the value for origin field with
`$objectsrc->element`. This make creation of invoice originating from a custom
module module impossible because we need the module part to instantiate
the class for the custom module object.

This also modify Common Object::fetch Object Linked() to append the
module part for `$sourcetype` and `$targettype` if it not set and object
have module property and it not in core module and 
`CommonObject::add_object_linked()` to append the module part if
object has `$module` property and is not in core modules.

indelog committed:
Fix creataing invoice originating for custom module

This fix the invoice create form when we provide `origin` element in
url. The actual create form overide the value for origin field with
`$objectsrc->element`. This make creation of invoice originating from a custom
module module impossible because we need the module part to instantiate
the class for the custom module object.

This also modify Common Object::fetch Object Linked() to append the
module part for `$sourcetype` and `$targettype` if it not set and object
have module property and it not in core module.

stickler-ci committed:
Fixing style errors.

indelog committed:
Fix Add module part in llx_element_element on CommonObject::add_object_linked()

Add module part to target type if object has $module property and is'nt
in core modules.

indelog committed:
Fix stickler

eldy commented on htdocs/compta/facture/card.php:
```diff
@@ -3730,8 +3730,10 @@ function setRadioForTypeOfIncoice() {
 		print '<input type="hidden" name="tva"      value="'.$objectsrc->total_tva.'">'."\n";
-		print '<input type="hidden" name="origin"   value="'.$objectsrc->element.'">';
-		print '<input type="hidden" name="originid" value="'.$objectsrc->id.'">';
+		// The lines below override the parameters set by GET or POST for origin value and also be present in $origin and $originid variables (it make creation of invoice originating from an external module object impossible because for get an external module class we also need to know the name of the module in addition of the name of the element).
```
If we remove this, we will lose them when submitting form and there is an error.
If origin and originid are provided on url, 
$objectsrc->id and $objectsrc->element should have the value of value provided.
So it overwrites yes, but with the same value to guarantee the data is propagated to the post.

indelog committed:
Add hook to overide sourcetype and targettype for linked object

This add hook `setLinkedObjectSourceTargetType` in
`CommonObject::showLinkToObjectBlock()` and in
`CommonObject::fetchObjectLinked()` to allow overiding sourcetype and
targettype by custom modules.

indelog commented:
I correct my first approach to customize sourcetype and targettype in methods
`CommonObject::showLinkToObjectBlock()` and CommonObject::fetchObjectLinked().
I thinked to use the `$coremodule` to determine if we must add the module
part to objecject element (wich is needed for get class path when the object
provide form a custom module).

It' a bad idea because it do not take care of other cusom modules and it make
the core code more complicated to maintain because we must take care of
updating the `$coremodule` value on each method which this variable is present
if we add `$module` property in object owned by per-existing core module.

Rather do this we can simply add a hook `setLinkedObjectSourceTargetType` for
allow developer of custom module to easily override sourcetype and targettype
if needed without impacting the core behavior.

_Example of hook usage :_

```
public function setLinkedObjectSourceTargetType($parameters, &$object, &$action, $hookmamager)
{
    if ($parameters['sourcetype'] == 'myobject') {
        $this->results['sourcetype'] = 'mymodule_myobject';
    }
    if ($parameters['targettype'] == 'myobject') {
        $this->results['targettype'] = 'mymodule_myobject';
    }
    return 1;
}
```

indelog committed:
Fix : Re-add bloc to set `$withtargettype` and `$withsourcetype`

eldy committed:
Update card.php
